File: verify/node/verify.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// #region check
// fields: decoded values; signField: "hash" for the launch context, "sign"
// for the getPhone envelope.
function checkSignature(fields, secret, signField) {
  const { [signField]: received = '', ...rest } = fields;
  const canon = Object.keys(rest).sort().map((k) => `${k}=${rest[k]}`).join('\n');
  const calc = crypto.createHmac('sha256', secret).update(canon, 'utf8').digest('hex');
  return calc.length === received.length &&
    crypto.timingSafeEqual(Buffer.from(calc), Buffer.from(received));
}
// #endregion check

const vector = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'vector.json'), 'utf8'));
let failed = false;
for (const c of vector.cases) {
  const ok = checkSignature(c.fields, vector.secret, c.sign_field);
  console.log(`${c.name}: ${ok ? 'ok' : 'FAIL'}`);
  if (!ok) failed = true;
}
process.exit(failed ? 1 : 0);

File: verify/python/verify.py
import hashlib
import hmac
import json
import sys
from pathlib import Path


# #region check
def check_signature(fields: dict, secret: str, sign_field: str) -> bool:
    """fields: decoded values; sign_field: "hash" for the launch context,
    "sign" for the getPhone envelope."""
    fields = dict(fields)
    received = fields.pop(sign_field, "")
    canon = "\n".join(f"{k}={fields[k]}" for k in sorted(fields))
    calc = hmac.new(secret.encode(), canon.encode(), hashlib.sha256).hexdigest()
    return hmac.compare_digest(calc, received)
# #endregion check


vector = json.loads((Path(__file__).parent / ".." / "vector.json").read_text())
failed = False
for case in vector["cases"]:
    ok = check_signature(case["fields"], vector["secret"], case["sign_field"])
    print(f"{case['name']}: {'ok' if ok else 'FAIL'}")
    failed = failed or not ok
sys.exit(1 if failed else 0)
